src/main/cpp/src/decimal_utils.cu

```diff
@@ -729,10 +746,11 @@ struct dec128_divider : public thrust::unary_function<cudf::size_type, __int128_
       auto const scale_divisor = pow_ten(n_shift_exp).as_128_bits();
 
       // The second divide gets the result into the scale that we care about and does the rounding.
-      auto const result = divide_and_round(first_div_result.quotient, scale_divisor);
+      auto const result = is_int_div ? integer_divide(first_div_result.quotient, scale_divisor)
+                            : divide_and_round(first_div_result.quotient, scale_divisor);
 
       overflows[i] = is_greater_than_decimal_38(result);
-      quotient_data[i] = result.as_128_bits();
+      quotient_data[i] = is_int_div ? result.as_64_bits() : result.as_128_bits();
```

Comment:
I have some problems with the code now that you are using templates and is_int_div.  We are mixing concerns, but we didn't fully define them.  Let me fix the problem with thrust::transform, and then we can talk about what the right way to fix this would be.

Reply:
Yeah using template classes is best for performance, as you can know all input types at compile time. Implementation would not be very flexible but you will get everything very clearly working.

The issue I see from `quotient_data[i] = is_int_div ? result.as_64_bits() : result.as_128_bits();` is that, `quotient_data[i]` is going to be assigned values of different types depending on `is_int_div`. This is prohibited in C++. If you really want to do that, `is_int_div` must be a template argument (i.e., compiled time constant).